File: BinarySearch.cpp
#include <algorithm>
#include <chrono>
#include <fstream>
#include <iostream>
#include <iterator>
#include <vector>

using std::cin, std::cout, std::endl;
using std::ifstream;
using std::istream_iterator;
using std::sort;
using std::vector;
using std::chrono::duration_cast;
using std::chrono::steady_clock;
using std::chrono::nanoseconds;

int rank(int key, const vector<int> &a) {
  for (int low = 0, high = a.size(); low < high;) {
    int mid = low + (high - low) / 2;
    if (key < a[mid]) {
      high = mid;
    } else if (key > a[mid]) {
      low = mid + 1;
    } else {
      return mid;
    }
  }
  return -1;
}

int main(int argc, char *argv[]) {
  steady_clock::time_point begin = steady_clock::now();
  ifstream fin(argv[1]);
  vector<int> whitelist((istream_iterator<int>(fin)), istream_iterator<int>());

  sort(whitelist.begin(), whitelist.end());

  for (int key; cin >> key;) {
    if (rank(key, whitelist) < 0) {
      cout << key << endl;
    }
  }
  steady_clock::time_point end = steady_clock::now();
  cout << "Time difference = "
       << duration_cast<nanoseconds>(end - begin).count() << "[ns]" << endl;
  // Time difference = 10054206292[ns]
}

File: BinarySearch.java

/**
 * BinarySearch
 */
import java.util.Arrays;

public class BinarySearch {

    public static int rank(int key, int[] a) {
        int low = 0;
        int high = a.length;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (key < a[mid]) {
                high = mid;
            } else if (key > a[mid]) {
                low = mid + 1;
            } else {
                return mid;
            }
        }
        return -1;
    }

    public static int countLessThan(int key, int[] a) {
        double k = key - 0.5;
        int low = 0;
        int high = a.length;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (k < a[mid]) {
                high = mid;
            } else if (k > a[mid]) {
                low = mid + 1;
            }
        }
        return high;
    }

    public static int count(int key, int[] a) {
        return countLessThan(key + 1, a) - countLessThan(key, a);
    }

    private static int[] readAllInts(String filename) {
        int[] input = new In(filename).readAllInts();
        Arrays.sort(input);
        int sz = 0;
        for (int i = 0; i < input.length; i++) {
            if (input[i] != input[sz]) {
                input[++sz] = input[i];
            }
        }
        ++sz;
        int[] whitelist = new int[sz];
        for (int i = 0; i < sz; i++) {
            whitelist[i] = input[i];
        }
        return whitelist;
    }

    public static void main(String[] args) {
        long startTime = System.nanoTime();
        //
        int[] whitelist = readAllInts(args[0]);
        //
        boolean blockMode = true;
        if (args.length > 1) {
            blockMode = args[1].equals("+");
            if (!blockMode && !args[1].equals("-")) {
                throw new IllegalArgumentException("Second argument must be '+' or '-'!");
            }
        }
        //
        while (!StdIn.isEmpty()) {
            int key = StdIn.readInt();
            boolean found = rank(key, whitelist) >= 0;
            if (blockMode && !found || !blockMode && found) {
                StdOut.println(key);
            }
        }
        //
        long endTime = System.nanoTime();
        StdOut.println("Time difference = " +  (endTime - startTime) + "[ns]");
        // Time difference = 17458320862[ns]
    }
}
